scrapers/faculty_directory_scraper/main.py

```diff
@@ -3,43 +3,126 @@
 
+        for key, cl in {
+            "biography": "bio",
+            "teaching": "teaching-summary",
+            "office-hours": "office-hours",
+            "current-teaching": "current-teaching",
+            "research": "research-summary",
+            "office": "location",
+            "website": "website",
+        }.items():
+            if tag := soup.find("div", f"field--name-field-{cl}"):
+                entry[key] = clean_string(tag.get_text(" "))
+
+        # Scraping the ORCID (see Gittens) is a bit more complicated
+        # because it isn't wrapped in a tag that is easily findable
+        if orcid_icon := soup.find("i", {"class": "fa-orcid"}):
+            entry["orcid"] = clean_string(orcid_icon.parent.text)
+
+        data[url] = entry
+
+
+async def entry_to_professor(session, url, faculty_list):
+    async with session.get(f"{BASE_URL}/{url}") as response:
+        soup = BeautifulSoup(await response.text("utf8"), "lxml")
+        canonical_url = (
+            "/" + soup.find("link", {"rel": "canonical"})["href"].split("/")[-1]
+        )
+
+        # If we are not at a search page, then we are at a faculty page.
+        # So add it to the list.
+        if canonical_url != "/search":
+            faculty_list.append(canonical_url)
+            return
+
+        # If we are at a search page, then we selected a name that has multiple professors.
+        # e.g. Michael Klein
+        # so we need to add both
+
+        # Getting each entry in the search results
+        for entry in soup.findAll("div", {"class": "views-field-title"}):
+            faculty_list.append(entry.find("a")["href"])
+
 
 async def main():
-    # Gets a json object that contains all faculty and their associated urls
-    faculty = requests.get("https://faculty.rpi.edu/data/peoplesearch").json()
-
-    data = {}
-    async with aiohttp.ClientSession() as s:
-        for professor in faculty["nodes"]:
-            await get_professor(
-                s,
-                professor,
-                f"https://faculty.rpi.edu{professor['node']['Path']}",
-                data,
-            )
+    # Gets the faculty directory page
+    response = requests.get(BASE_URL)
+    faculty_entries = html.fromstring(
+        response.content.decode(response.apparent_encoding)
+    ).xpath('//select[@data-drupal-facet-id="name"]/option/@value')
+
+    async with aiohttp.ClientSession() as session:
+        # when two professors have the same name, the faculty_entries variable
+        # only has one entry for both, so we need to account for this
+        faculty_list = []
+        for faculty_url in faculty_entries:
+            await entry_to_professor(session, faculty_url, faculty_list)
+        # now that we actually have a list of professors, we will fill in the
+        # data object with the relevant information
+        data = {}
+        for professor_url in faculty_list:
+            await get_professor(session, professor_url, data)
```

Comment:
I dont really care that much but I am pretty sure for ever unique named professor you are requesting their page twice. Once in the `entry_to_professor` function and then again here. That could be simplified to reduce load on the faculty website by about half and probably speed up the code by almost 2x. But it is not a huge deal if it is a pain to fix that.

Reply:
Yeah I noticed that after I finished writing the code. I am not sure how to restructure this to avoid that, but I'll try to.

Reply:
Addressed in latest commit